### ISSUE TITLE
OpenSSL configuration leads to build failure in otp_test_engine.c

### ISSUE THREAD

**Issue by erwan-le-pape** [erlang/otp#4573]
**Describe the bug**
In `lib/crypto/c_src/otp_test_engine.c`, function prototypes are declared in a twice-conditional block that depends on `HAVE_EC` (from l.58 to l.378) and `FAKE_RSA_IMPL`. Functions are however are declared on the sole `FAKE_RSA_IMPL` condition (from l.380 to l.456). This breaks compilation on OpenSSL that matches conditions for `HAVE_EC` to be false with typical hardening option `-Werror=format-security`.

**To Reproduce**
Compiling on an OpenSSL 1.1.1i with `OPENSSL_NO_ECDSA` and `-Werror=format-security` in `CFLAGS` works for me, I expect any OpenSSL >= 0.9.8o and with any of `OPENSSL_NO_EC`, `OPENSSL_NO_ECDH` or `OPENSSL_NO_ECSDA` would do the trick.

**Expected behavior**
The function implementations should likely be within the `#if defined(HAVE_EC)` section, in addition to which, I'm unsure why the `HAVE_EC` check looks beyond `OPENSSL_NO_EC` since it doesn't actually depend on `OPENSSL_NO_ECDSA` (it compiles just fine without checking for OPENSSL_NO_ECDSA on OpenSSL 1.1.1i and OTP 23.2.7, although I can't confirm backwards compatibility with obsolete OpenSSL <= 1.1.0).

**Affected versions**
20.3.8.15 through 23.2.7 (verified on 23.2.6 & 23.2.7).

**Additional context**
Compiling fails because of the `implicit declaration of function` but I've left the information the entire file.
```
otp_test_engine.c:387:5: warning: no previous prototype for ‘test_rsa_sign’ [-Wmissing-prototypes]
 int test_rsa_sign(int dtype,
     ^
otp_test_engine.c: In function ‘test_rsa_sign’:
otp_test_engine.c:416:9: error: implicit declaration of function ‘add_test_data’ [-Werror=implicit-function-declaration]
         add_test_data(sigret, (unsigned int)slen); /* The signature is 0,1,2...255,0,1... */
         ^
otp_test_engine.c: At top level:
otp_test_engine.c:426:5: warning: no previous prototype for ‘test_rsa_verify’ [-Wmissing-prototypes]
 int test_rsa_verify(int dtype,
     ^
otp_test_engine.c: In function ‘test_rsa_verify’:
otp_test_engine.c:445:13: error: implicit declaration of function ‘chk_test_data’ [-Werror=implicit-function-declaration]
             && chk_test_data(sigret, siglen);
             ^
otp_test_engine.c: At top level:
otp_test_engine.c:450:12: warning: ‘test_rsa_free’ defined but not used [-Wunused-function]
 static int test_rsa_free(RSA *rsa)
            ^
```

**Proposed patch**
```patch
--- lib/crypto/c_src/otp_test_engine.c       2020-11-30 09:43:07.000000000 +0000
+++ lib/crypto/c_src/otp_test_engine.c       2020-11-30 09:43:07.000000000 +0000
@@ -375,8 +375,6 @@
     return 0;
 }

-#endif
-
 #if defined(FAKE_RSA_IMPL)
 /* RSA sign. This returns a fixed string so the test case can test that it was called
    instead of the cryptolib default RSA sign */
@@ -454,3 +452,5 @@
 }

 #endif /* if defined(FAKE_RSA_IMPL) */
+
+#endif /* if defined(HAVE_EC) */
```

**Comment by HansN** [comment]
We already have an internal ticket about this, but it is not prioritized.
Since you kindly proposes a patch, would you mind submiting it as a [Pull Request](https://github.com/erlang/otp/blob/master/CONTRIBUTING.md)?  In this special case no documentation or test case is needed.

**Comment by erwan-le-pape** [comment]
Sure, just did :)

**Comment by HansN** [comment]
Fixed in OTP-23.3
